src/main/python/deep_qa/layers/entailment_models/tuple_match.py

```diff
@@ -0,0 +1,180 @@
+            self.trainable_weights.extend(hidden.trainable_weights)
+
+        super(TupleMatch, self).build(input_shape)
+
+    @overrides
+    def compute_mask(self, inputs, input_mask=None):  # pylint: disable=unused-argument
+        # Here, input_mask is optionally a list of two masks (corresponding to the list of inputs) or None.
+        # Each of the masks in input_mask has shape: (batch size, num_slots)
+        # We want to see if either of the inputs is all padding (i.e. the mask would be all 0s), if so, then the
+        # whole tuple_match should be masked, so we would return a 0, otherwise we return a 1.  As such, the
+        # shape of the returned mask is (batch size, 1)
+        if any(input_mask):
+            mask1 = input_mask[0]
+            mask2 = input_mask[1]
+            mask1 = K.cast(K.any(mask1), 'int32')
+            mask2 = K.cast(K.any(mask2), 'int32')
+            return K.cast((mask1 + mask2) >= 2, 'int32')
+
+        return None
+
+    def call(self, inputs, mask=None):
+        tuple1_input = inputs[0]    # shape: (batch size, num_slots, num_slot_words_t1)
+        tuple2_input = inputs[1]    # shape: (batch size, num_slots, num_slot_words_t2)
+        # Check that the tuples have the same number of slots.
+        assert K.int_shape(tuple1_input)[1] == K.int_shape(tuple2_input)[1]
+
+        # Expand tuple1 to shape: (batch size, num_slots, num_slot_words_t1, num_slot_words_t2)
+        expanded_tuple1 = K.expand_dims(tuple1_input, 3)    # now (b, num_slots, num_slot_words_tuple1, __ )
+        tiled_tuple1 = K.repeat_elements(expanded_tuple1, K.int_shape(tuple2_input)[2], axis=3)
+
+        # Expand tuple2 to shape: (batch size, num_slots, num_slot_words_t1, num_slot_words_t2)
+        expanded_tuple2 = K.expand_dims(tuple2_input, 2)    # now (b, num_slots, __, num_slot_words_tuple2)
+        tiled_tuple2 = K.repeat_elements(expanded_tuple2, K.int_shape(tuple1_input)[2], axis=2)
+
+        # This generates a binary tensor of the same shape as tiled_tuple1 /
+        # tiled_tuple2 that indicates if given word matches between tuple1 and tuple2 in a particular slot.
+        # Currently, we only consider S_t1 <--> S_t2 etc overlap, not across slot types.
+        # shape: (batch size, num_slots, numslot_words_tuple1, num_slot_words_tuple2)
+        tuple_words_overlap = K.cast(K.equal(tiled_tuple1, tiled_tuple2), "float32")
+
+        # Exclude zeros (i.e. padded elements) from matching each other.
+        # tuple1_mask is 1 if tuple1 has a real element, 0 if it's a padding element.
+        tuple1_mask = K.cast(K.not_equal(tiled_tuple1, K.zeros_like(tiled_tuple1, dtype='float32')),
+                             dtype='float32')
+        zeros_excluded_overlap = tuple_words_overlap * tuple1_mask
+
+        # Find non-padding elememts in tuple1.
+        # shape: (batch size, num_slots, num_slot_words_tuple1)
+        non_padded_tuple1 = K.cast(K.not_equal(tuple1_input, K.zeros_like(tuple1_input)), 'float32')
```

Comment:
Oh, this is the non-tiled version - you should probably put `tiled` in the name of `tuple1_mask`, so that's more clear.